### ISSUE TITLE
Deliver an explicit opt-in GB10/cloud recognition backend with the same demo contract

### ISSUE THREAD

**Issue by nino96** [nino96/chess-ocr#4]
## Outcome
An OPTIONAL local-server backend (GB10 first; deployable in an owner-approved cloud environment) provides the same geometry/placement/uncertainty contract as the browser recognizer. Browser offline mode remains functional and never silently uploads.

Depends on merged #1 contract and #2 accepted data. Coordinate with #3; full offline success is NOT a prerequisite if its measured limits justify the fallback. Implement a local service first. Cloud account creation, provisioning, spending, external exposure and private-data uploads require separate explicit approval.

## Model and service
- Running the same small model remotely can improve latency but cannot improve its predictions. Demonstrate this transparently when first validating the service transport.
- Quality-oriented candidate: native pretrained RF-DETR Small, after exact checkpoint/revision/license/training-data provenance review. Core model license lane is distinct from XL/2XL variants; do not substitute model variants silently. Verify Linux ARM64/GB10 execution and export with the actual supported stack.
- Train at higher input resolution for board and occupied-cell recognition, using #2 common page/grid labels. Full-page board localization and crop recognition may be staged. Define occupied-cell region targets honestly; grid cells are not automatically tight glyph bounding boxes. Map centers to verified grid with duplicate/missing-cell and orientation uncertainty handling.
- Freeze preprocessing, training schedule, per-run budgets, confidence and candidate selection before training, under a separately allocated shared ledger. Reuse label/geometry/eval tools; no forked ground-truth dataset. Do not train the fallback just to exhaust the budget.
- Evaluate all baseline/offline/server candidates on identical raw inputs, with end-to-end localization and confidently-wrong/clean-preservation metrics. #3 quality gates apply to any claimed qualified backend. Keep test exposure tracked across candidates; do not repeatedly tune against qualification.
- Provide a small Python HTTP adapter with versioned schema, health/readiness, bounded image dimensions/decoded pixels/bytes, timeouts, cancellation/backpressure and validated responses. Store native weights outside Git; provide explicit hash-bound acquisition.
- Container/installation docs for Linux ARM64 GB10; document x86_64 cloud differences. Do not promise untested ARM64/TensorRT compatibility. Native CPU/server smoke must be available where practical.

## ChessQueries: owner-added evaluation candidate (2026-09-06)

Add ChessQueries alongside RF-DETR Small as a **candidate for bounded local GB10 evaluation**, not a selected replacement or a qualified recognizer. Independent setup/research may precede #1/#2 integration; service delivery still depends on their merged contract and accepted data. Keep this work in #4. The offline WASM default and #3 gates remain unchanged.

The architecture uses a DINOv2 ViT-L/14 encoder plus 64 learned square queries and a DETR-style decoder, producing 13-class square predictions. Its physical-board/broadcast benchmarks do not establish printed-page recognition. Compare raw page/selection inputs separately from exact crops, report geometry/multiple-board limitations, orientation uncertainty, confident errors and baseline-correct losses. Do not infer image-relative geometry from attention maps or invent unobservable FEN state.

### Pinned inputs and review before acquisition

- [Paper](https://arxiv.org/html/2608.30762v1).
- [Source revision](https://github.com/JSeytre/chessqueries/tree/2a9f3c00619133541c65de4315149037b5dece80): `2a9f3c00619133541c65de4315149037b5dece80`.
- [Model revision](https://huggingface.co/joelseytre/chessqueries/tree/4972a833d62e6d16f686f021608921c49e1d180c): `4972a833d62e6d16f686f021608921c49e1d180c`.
- Tensor-only artifact: `chessqueries-vitL14-644-joint.safetensors`, 1,486,661,004 bytes; SHA-256 `6151bdd98fbe25f32080c097eba7ae75808b5615160e4867240931085fc762e5`.
- Code and weights list PolyForm Noncommercial 1.0.0. Dataset annotations have separate terms. Before downloading/installing, record artifact-specific local-use rights/privacy review, dated license evidence/hashes, dependency provenance and a reservation in the shared project ledger. This candidate addition does not approve redistribution, publication, private inputs or training. Unknown pretraining overlap stays unknown.
- No original `.ckpt`/pickle loading. Use the verified safetensors file and explicit checkpoint path. Do not run upstream annotation, dataset-download, LLM or shared-demo workflows.

### GB10 setup recipe — proposed, not yet executed or validated

Reference machine prerequisite check: Ubuntu 24.04 ARM64, Python 3.12.3, NVIDIA GB10, driver 580.159.03. `nvidia-smi` worked outside the agent sandbox; its displayed CUDA version is driver capability, not proof of a working PyTorch stack.

Upstream's lock selects torch 2.8.0 / torchvision 0.23.0 and does not establish GB10 GPU support. NVIDIA's [Spark dependency guide](https://docs.nvidia.com/dgx/dgx-spark-porting-guide/porting/dependencies.html) lists CUDA 13.0. The recipe below deliberately creates a **local experimental lock**, using the official CUDA 13.0 index and torch 2.9.1 / torchvision 0.24.1. This is an explicit upstream dependency deviation requiring native inference/parity checks, not an upstream-tested setup. No global Python installs.

Official CPython 3.12 ARM64 wheel identities checked before drafting:

- torch `2.9.1+cu130`: SHA-256 `6e7f84cb10c7e7d9f862c318f056d64840544ab4f0bcbf8cf7ed6047fe04051f` ([index](https://download.pytorch.org/whl/cu130/torch/)).
- torchvision `0.24.1`: SHA-256 `89743dcee13e943f58b37c7647aff14b5bb24c11c84826376d457acf97586fec` ([index](https://download.pytorch.org/whl/cu130/torchvision/)).

After the review and resource reservation, from the chess-ocr root, use a fresh ignored checkout:

```sh
set -eu
uname -m
nvidia-smi
mkdir -p work/chessqueries
git clone https://github.com/JSeytre/chessqueries.git work/chessqueries/source
cd work/chessqueries/source
git checkout --detach 2a9f3c00619133541c65de4315149037b5dece80
python3.12 -m venv .venv
. .venv/bin/activate
export PIP_CACHE_DIR="$PWD/../pip-cache"
export POETRY_CACHE_DIR="$PWD/../poetry-cache"
export HF_HOME="$PWD/../hf-cache"
export TORCH_HOME="$PWD/../torch-cache"
python -m pip install 'poetry==2.2.1'
poetry source add --priority=explicit pytorch-cu130 https://download.pytorch.org/whl/cu130
poetry add --lock --source pytorch-cu130 'torch==2.9.1+cu130' 'torchvision==0.24.1'
# Inspect the resolved ARM64 artifacts/hashes and local manifest/lock diff before install.
git diff -- pyproject.toml poetry.lock
poetry install --only main
python -m pip check
mkdir -p ../evidence
python -m pip freeze > ../evidence/installed.txt
sha256sum pyproject.toml poetry.lock > ../evidence/environment.sha256
```

Stop on resolution, binary-wheel, import or CUDA failure and report the exact blocker. Do not silently use CPU, upgrade more packages, compile a replacement stack or try a series of alternate stacks. Preserve the resolved lock and package provenance locally; promote a reviewed platform lock only after validation. Shell setup is a one-time recipe; subsequent sessions activate the existing venv rather than reclone/reinstall.

Verify actual CUDA execution before acquiring the model:

```sh
python - <<'PY'
import platform, sys, torch, torchvision
assert sys.prefix != sys.base_prefix
assert platform.machine() == 'aarch64'
assert torch.cuda.is_available(), 'CUDA unavailable: stop, no silent CPU fallback'
print('torch', torch.__version__, 'torchvision', torchvision.__version__)
print('CUDA runtime', torch.version.cuda)
print('device', torch.cuda.get_device_name(0))
print('capability', torch.cuda.get_device_capability(0))
print('compiled architectures', torch.cuda.get_arch_list())
a = torch.arange(256, dtype=torch.float32, device='cuda').reshape(16, 16)
b = a @ a.T
torch.cuda.synchronize()
torch.testing.assert_close(b.cpu(), a.cpu() @ a.cpu().T)
print('CUDA matrix operation passed; model compatibility still unverified')
PY
```

Acquire only the reviewed immutable model artifact, verify its bytes, then use an explicitly approved local diagnostic image (never a private book by default):

```sh
mkdir -p checkpoints/release
curl --fail --location --retry 2 --max-time 1800 \
  'https://huggingface.co/joelseytre/chessqueries/resolve/4972a833d62e6d16f686f021608921c49e1d180c/chessqueries-vitL14-644-joint.safetensors' \
  --output checkpoints/release/chessqueries-vitL14-644-joint.safetensors.part
printf '%s  %s\n' \
  '6151bdd98fbe25f32080c097eba7ae75808b5615160e4867240931085fc762e5' \
  'checkpoints/release/chessqueries-vitL14-644-joint.safetensors.part' | sha256sum --check
mv checkpoints/release/chessqueries-vitL14-644-joint.safetensors.part \
  checkpoints/release/chessqueries-vitL14-644-joint.safetensors
# Set CQ_INPUT to one approved local image; output can contain its identity and FEN.
: "${CQ_INPUT:?Set CQ_INPUT to an approved local diagnostic image}"
timeout 300s poetry run chessqueries-predict "$CQ_INPUT" \
  --device cuda --batch-size 1 --resolution 644 \
  --checkpoint checkpoints/release/chessqueries-vitL14-644-joint.safetensors \
  > ../evidence/inference-local.txt 2>&1
```

Do not publish that raw output: upstream CLI can include input paths and positions. Keep it in ignored local storage. Use the CLI rather than a web demo for this screening. All weights, caches, modified upstream manifests and generated results remain under ignored `work/`. Inspect whether inference attempts additional network access; cache/review any required asset explicitly and demonstrate offline inference after readiness.

### Bounded evaluation and completion evidence

Before execution, freeze one screening set of at most 20 approved printed-diagram development inputs spanning new glyph/page conditions, plus identical baseline inputs/hashes. Declare concrete download/storage/CPU/GPU/review ceilings in the project ledger. Begin with one setup attempt, one CUDA smoke and one unchanged-checkpoint screen; no fine-tuning, architecture variants or automatic seed sweeps are authorized by this addition. One infrastructure failure permits a bounded diagnosis; any replacement stack or training proposal needs a measured reason and separately reserved budget. Use a resumable job/status artifact for lengthy work; no AI log polling.

- [ ] Resolve and preserve a hash-bound ARM64 environment; record source/model/license/dependency hashes and actual commands/device.
- [ ] Demonstrate real CUDA operation and full model inference; CPU/GPU placement/probability parity on a small fixed approved input set, with declared tolerances and discrepancies recorded.
- [ ] Screen unchanged weights on printed-diagram development inputs, reporting all-input exact-board accuracy, per-class/occupied/color errors, uncertainty and limitations. This small screen is not qualification.
- [ ] Record page/selection localization, grid geometry, orientation and multiple-board gaps separately; exact-crop success is not end-to-end delivery.
- [ ] Record functional CUDA/model inference evidence and input coverage; record failed attempts, not selected successes. Device-level latency, memory, and deployment-capacity qualification belong to the consuming integration environment.
- [ ] Record advance/defer/reject decision with evidence. ONNX export, browser WASM parity and full server integration remain separate unrun gates until reached; no browser-feasibility claim from GB10 timings.

## Privacy and security
- Default backend is browser/offline. Remote mode requires deliberate endpoint selection and explicit upload consent identifying what pixels go where. No silent fallback, automatic whole-book upload, telemetry, raw request logs, FEN/image/path retention or third-party forwarding.
- Bind loopback by default. Remote exposure requires authentication, TLS at deployment boundary, explicit CORS origins, request/rate/concurrency limits and endpoint/config validation. Browser must not forward credentials to a different origin; never put secrets into public bundles.
- No request-body/image caching. Document in-memory lifetime and response/error behavior. User-confirmed edits survive retries, response reordering and backend switching.
- Provide opt-out and clear mode/endpoint indication in the same demo. Offline and consent-rejection tests must prove no request is sent.

## Acceptance
- [ ] Demo can deliberately choose browser or configured server; identical response contract and editable output, with honest backend/model identity.
- [ ] Actual GB10 inference and optional larger-model comparison are evidenced or hardware/model limitations recorded without claiming success.
- [ ] Auth/CORS/input limits/timeout/cancellation/backpressure/stale-response/no-retention and no-upload-without-consent tests pass.
- [ ] Reproducible local model/service setup and versioned manifests delivered; cloud guidance does not imply cloud deployment was performed. Device-level latency, memory, and concurrency qualification are not gates in this repository.
- [ ] Any accuracy superiority is established on frozen representative inputs, not inferred from GPU hardware/model size.
- [ ] No downloaded assets/binaries/secrets in Git; notices and unresolved distribution obligations explicit.

Keep service/model increments here, not one issue per deployment or seed.

**Comment by nino96** [comment]
Reopened immediately. The previous closure over-interpreted the owner’s statement about excluding memory, iPad, and similar application-qualification work. The owner did not authorize removing the GB10/server model workstream; #4 remains active pending an explicit scope decision.
